File: src/test/scala/io/aif/associations/builder/ExperimentsConnectionsGraphBuilderTest.scala
package io.aif.associations.builder

import org.scalatest.FunSuite
import org.scalatest.BeforeAndAfter


class ExperimentsConnectionsGraphBuilderTest extends FunSuite with BeforeAndAfter {

  var experimentsConnectionsGraphBuilder: ExperimentsConnectionsGraphBuilder[Int] = _

  before {
    experimentsConnectionsGraphBuilder = new ExperimentsConnectionsGraphBuilder[Int](exp => 1)
  }

  test("verifying the build process on empty list") {
    assert (experimentsConnectionsGraphBuilder build List[Int]() isEmpty)
  }

  test("verifying the build process with one element") {
    val result = experimentsConnectionsGraphBuilder build 1 :: Nil
    assert (result.size == 1)
  }

  test("verifying the build process with two same elements") {
    val result = experimentsConnectionsGraphBuilder build 1 :: 1 :: Nil
    assert (result.size == 1)
    assert (result.get(1).get isEmpty)
  }

  test("verifying the build process with two different elements") {
    val result = experimentsConnectionsGraphBuilder build 1 :: 2 :: Nil
    assert (result.size == 2)
    assert (result.get(1).get.get(2).get(0) == 2.0)
    assert (result.get(2).get isEmpty)
  }

  test("verifying the build process with three elements") {
    val result = experimentsConnectionsGraphBuilder build 1 :: 2 :: 2 :: Nil
    assert (result.size == 2)
    assert (result.get(1).get.get(2).get(0) == 2.0)
    assert (result.get(1).get.get(2).get(1) == 3.0)
    assert (result.get(2).get isEmpty)
  }

  test("verifying the build process with three elements and non default multilayer") {
    experimentsConnectionsGraphBuilder = new ExperimentsConnectionsGraphBuilder[Int](exp => if (exp == 1) 1.0 else 2.0)
    val result = experimentsConnectionsGraphBuilder build 1 :: 2 :: 2 :: Nil
    assert (result.size == 2)
    assert (result.get(1).get.get(2).get(0) == 4.0)
    assert (result.get(1).get.get(2).get(1) == 10.0)
    assert (result.get(2).get isEmpty)
  }

  test("verifying the build process with fore elements and non default multilayer") {
    experimentsConnectionsGraphBuilder = new ExperimentsConnectionsGraphBuilder[Int](exp => if (exp == 1) 1.0 else 2.0)
    val result = experimentsConnectionsGraphBuilder build 1 :: 2 :: 2 :: 2 :: Nil
    assert (result.size == 2)
    assert (result.get(1).get.get(2).get(0) == 4.0)
    assert (result.get(1).get.get(2).get(1) == 10.0)
    assert (result.get(2).get isEmpty)
  }
  
}

File: src/test/java/io/aif/associations/builder/ExperimentsConnectionsGraphBuilderTest.java
package io.aif.associations.builder;

import io.aif.associations.model.IGraph;
import org.junit.Test;

import java.util.*;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class ExperimentsConnectionsGraphBuilderTest {

    @Test
    public void testBuildWithEmptyExperimentsList() throws Exception {
        final List<Integer> experiments = Collections.emptyList();

        final ExperimentsConnectionsGraphBuilder<Integer> experimentsConnectionsGraphBuilder = new ExperimentsConnectionsGraphBuilder<>();

        final Map<Integer, Map<Integer, Double>> result = experimentsConnectionsGraphBuilder.build(experiments);

        assertTrue(result.isEmpty());
    }

    @Test
    public void testBuildWithOneExperiment() throws Exception {
        final List<Integer> experiments = Arrays.asList(1);

        final ExperimentsConnectionsGraphBuilder<Integer> experimentsConnectionsGraphBuilder = new ExperimentsConnectionsGraphBuilder<>();

        final Map<Integer, Map<Integer, Double>>  result = experimentsConnectionsGraphBuilder.build(experiments);

        final Integer element = result.keySet().iterator().next();

        assertEquals(result.keySet().size(), 1);
        assertTrue(result.get(element).keySet().isEmpty());
    }

    @Test
    public void testBuildWithTwoSameExperiments() throws Exception {
        final List<Integer> experiments = Arrays.asList(1, 1);

        final ExperimentsConnectionsGraphBuilder<Integer> experimentsConnectionsGraphBuilder = new ExperimentsConnectionsGraphBuilder<>();

        final Map<Integer, Map<Integer, Double>>  result = experimentsConnectionsGraphBuilder.build(experiments);

        final Integer element = result.keySet().iterator().next();

        assertEquals(result.keySet().size(), 1);
        assertTrue(result.get(element).keySet().isEmpty());
    }

    @Test
    public void testBuildWithTwoSameExperimentsAmdOneNotSameAtTheEnd() throws Exception {
        final List<Integer> experiments = Arrays.asList(1, 1, 2);

        final ExperimentsConnectionsGraphBuilder<Integer> experimentsConnectionsGraphBuilder = new ExperimentsConnectionsGraphBuilder<>();

        final Map<Integer, Map<Integer, Double>>  result = experimentsConnectionsGraphBuilder.build(experiments);

        final Integer element1 = result.keySet().stream().filter(i -> i == 1).findFirst().get();
        final Integer element2 = result.keySet().stream().filter(i -> i == 2).findFirst().get();

        assertEquals(result.get(element1).get(element2), (Double)1.0);
        assertEquals(result.get(element1).keySet().size(), 1);
        assertEquals(result.get(element2).keySet().size(), 0);
    }

    @Test
    public void testBuildWithTwoSameExperimentsAmdOneNotSameAtTheBegining() throws Exception {
        final List<Integer> experiments = Arrays.asList(2, 1, 1);

        final ExperimentsConnectionsGraphBuilder<Integer> experimentsConnectionsGraphBuilder = new ExperimentsConnectionsGraphBuilder<>();

        final Map<Integer, Map<Integer, Double>>  result = experimentsConnectionsGraphBuilder.build(experiments);

        final Integer element1 = result.keySet().stream().filter(i -> i == 1).findFirst().get();
        final Integer element2 = result.keySet().stream().filter(i -> i == 2).findFirst().get();

        assertTrue(result.get(element1).keySet().isEmpty());
        assertEquals(result.get(element2).get(element1), (Double)3.0);
    }

    @Test
    public void testBuildWithTwoSameExperimentsAmdOneNotSameAtTheBeginingWithNonDefaultMultiplexer() throws Exception {
        final List<Integer> experiments = Arrays.asList(2, 1, 1);

        final ExperimentsConnectionsGraphBuilder<Integer> experimentsConnectionsGraphBuilder = new ExperimentsConnectionsGraphBuilder<>(d -> d == 1 ? 4. : 1.);

        final Map<Integer, Map<Integer, Double>>  result = experimentsConnectionsGraphBuilder.build(experiments);

        final Integer element1 = result.keySet().stream().filter(i -> i == 1).findFirst().get();
        final Integer element2 = result.keySet().stream().filter(i -> i == 2).findFirst().get();

        assertTrue(result.get(element1).keySet().isEmpty());
        assertEquals(result.get(element2).get(element1), (Double)11.0);
    }

}